cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/output/MultipleOutputsMainOutputWrapper.java

```diff
@@ -49,6 +49,14 @@
   @Override
   public RecordWriter<K, V> getRecordWriter(TaskAttemptContext job) throws IOException, InterruptedException {
     OutputFormat<K, V> rootOutputFormat = getRootOutputFormat(job);
+    List<String> namedOutputsList = MultipleOutputs.getNamedOutputsList(job);
+    if (namedOutputsList != null && namedOutputsList.size() == 1) {
```

Comment:
It's confusing what is happening here from just reading this code. 

Is the namedOutputsList always non-empty? If it's empty it means the bug still exists right?

Is this RecordWriter only used in the case where there's one output? Can you add javadocs to describe the contract?

Reply:
on further reading of the code, it seems like the "root" output format is the output format if one CDAP output is added, or the output format if no CDAP outputs were added, or some unsupported output format if more than one CDAP output was added. 

This is super confusing because none of this logic is in this class -- it's spread across MultipleOutputs and MapReduceRuntimeService. Let's clean this up by removing MultipleOutputs.addNamedOutput() and replacing it with a static method in this class

static void addOutputs(List<ProvidedOutput> providedOutputs);

The method should basically do what MapReduceRuntimeService.setOutputsIfNeeded() does, except it will do it in this class, where it will be more clear what the root is and what is contained in the named outputs.

Reply:
namedOutputsList should also never be null, you don't need to check for that